test/mpi/dtpools/src/dtpools_internal.c

```diff
@@ -0,0 +1,1167 @@
+/* Datatype Pool Object Create Functions                     */
+/* --------------------------------------------------------- */
+
+int DTPI_Struct_create(struct params *par, DTP_t dtp)
+{
+    int i, err = DTP_SUCCESS;
+    int basic_type_count = 0;
+    int *basic_type_sizes = NULL;
+    int *basic_type_counts;
+    int len, size = 0;
+    MPI_Aint displs, *basic_type_displs = NULL;
+    MPI_Datatype *basic_types;
+    void *buf;
+    DTPI_t *dtpi = NULL;
+    char **basic_type_names = NULL;
+    char type_name[TYPE_NAME_MAXLEN] = { 0 };
+
+    /* allocate space for private datatype info */
+    dtpi = (DTPI_t *) malloc(sizeof(*dtpi));
+    if (!dtpi) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+
+    basic_type_count = dtp->DTP_type_signature.DTP_pool_struct.basic_type_count;
+    basic_type_names = (char **) malloc(sizeof(*basic_type_names) * basic_type_count);
+    basic_type_sizes = (int *) malloc(sizeof(*basic_type_sizes) * basic_type_count);
+    basic_type_displs = (MPI_Aint *) malloc(sizeof(*basic_type_displs) * basic_type_count);
+    if (!basic_type_names || !basic_type_sizes || !basic_type_displs) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+
+    for (i = 0; i < basic_type_count; i++) {
+        basic_type_names[i] = (char *) malloc(sizeof(**basic_type_names) * BASIC_TYPE_NAME_MAXLEN);
+        if (!basic_type_names[i]) {
+            err = DTP_ERR_OTHER;
+            fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+            fflush(stdout);
+            goto fn_fail;
+        }
+    }
+
+    /* compute struct size */
+    basic_type_counts = dtp->DTP_type_signature.DTP_pool_struct.DTP_basic_type_count;
+    basic_types = dtp->DTP_type_signature.DTP_pool_struct.DTP_basic_type;
+    for (i = 0; i < basic_type_count; i++) {
+        err = MPI_Type_size(basic_types[i], &basic_type_sizes[i]);
+        if (err) {
+            DTPI_Print_error(err);
+        }
+        err = MPI_Type_get_name(basic_types[i], basic_type_names[i], &len);
+        if (err) {
+            DTPI_Print_error(err);
+        }
+        size += (basic_type_sizes[i] * basic_type_counts[i]);
+    }
+
+    /* compute displacements */
+    displs = 0;
+    for (i = 0; i < basic_type_count; i++) {
+        basic_type_displs[i] = displs;
+        displs += basic_type_sizes[i] * basic_type_counts[i];
+    }
+
+    /* allocate buffer */
+    buf = (void *) malloc(size);
+    if (!buf) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+    memset(buf, 0, size);
+
+    /* create struct datatype */
+    err = MPI_Type_struct(basic_type_count, basic_type_counts, basic_type_displs, basic_types,
```

Comment:
This function is already deprecated, -> `MPI_Type_create_struct`.

Reply:
Ok I didn't know that I will update with the appropriate function